File: Minimum Multiplications to reach End - GFG/minimum-multiplications-to-reach-end.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int minimumMultiplications(vector<int>& arr, int start, int end) {
        queue<pair<int,int>> q;
        vector<int> dist(1e5,1e9);
        dist[start]  =0;
        q.push({start,0});
        int mod = 1e5;
        //TIME COMPLEXITY
        // O(1e5 * N)
        while(!q.empty()) {
            int node = q.front().first;
            int steps = q.front().second;
            q.pop();
            if(node == end) return steps;
            for(auto it : arr) {
                int num = (it * node) % mod;
                if(steps + 1 < dist[num]) {
                    dist[num] = steps + 1;
                    
                    q.push({num,steps + 1});
                }
            }
        }
        return -1;
    }
};


//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> arr(n);
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        int start, end;
        cin >> start >> end;
        Solution obj;
        cout << obj.minimumMultiplications(arr, start, end) << endl;
    }
}

// } Driver Code Ends

File: Minimum Multiplications to reach End - GFG/minimum-multiplications-to-reach-end.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

// Position this line where user code will be pasted.

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            int n = sc.nextInt();
            int[] a = new int[n];

            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
            }

            int start = sc.nextInt();
            int end = sc.nextInt();

            Solution ob = new Solution();
            int ans = ob.minimumMultiplications(a, start, end);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


// User function Template for Java
class Pair{
    int src;
    int dist;
    Pair(int src,int dist) {
        this.src = src;
        this.dist = dist;
    }
}
class Solution {
    int minimumMultiplications(int[] arr, int start, int end) {
        int mod =  100000;
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(start ,0)) ;
        int dist[] = new int[100000];
        Arrays.fill(dist,(int)1e9);
        dist[start] = 0;
        while(!q.isEmpty()) {
            int node = q.peek().src;
            int steps = q.peek().dist;
            q.remove();
            for(int it : arr) {
                int num = (it * node) % mod;
                
                if(steps + 1 < dist[num]) {
                    dist[num] = steps + 1;
                    if(num == end ) return steps + 1;
                    q.add(new Pair(num, steps + 1));
                }
            }
        }
        return -1;
    }
}
